File: backend/app/modules/market/services/asset_health.py
import uuid
from datetime import datetime, timezone
from typing import Any

from app.core.config import settings
from app.core.redis import get_cached_asset_signals
from app.modules.market.entities.market import AssetHealth
from app.core.services.base import BaseService
from app.modules.market.repositories.asset_fundamentals import AssetFundamentalsRepository
from app.modules.market.repositories.asset_health import AssetHealthRepository
from app.modules.market.repositories.market import MarketRepository


def _evaluate_quote_sla(quote_age_seconds: int | None) -> bool:
    return quote_age_seconds is not None and quote_age_seconds <= settings.SLA_QUOTE_MAX_AGE_SEC


def _evaluate_fundamentals_sla(fundamentals_age_seconds: int | None) -> bool:
    return fundamentals_age_seconds is not None and fundamentals_age_seconds <= settings.SLA_FUNDAMENTALS_MAX_AGE_SEC


def _evaluate_news_sla(news_age_seconds: int | None) -> bool:
    return news_age_seconds is not None and news_age_seconds <= settings.SLA_NEWS_MAX_AGE_SEC


def _evaluate_signal_sla(signal_age_seconds: int | None) -> bool:
    return signal_age_seconds is not None and signal_age_seconds <= settings.SLA_SIGNAL_MAX_AGE_SEC


def _dimension_status(age_seconds: int | None, healthy: bool) -> str:
    """Tri-state label for a single health dimension — 'unknown' (no data to
    evaluate) is distinct from 'healthy'/'unhealthy' (data was evaluated)."""
    if age_seconds is None:
        return "unknown"
    return "healthy" if healthy else "unhealthy"


def evaluate_health_status(quote_age: int | None, news_age: int | None, signal_age: int | None) -> str:
    # Allowed values: HEALTHY, STALE, DEGRADED, UNKNOWN
    if quote_age is None and news_age is None and signal_age is None:
        return "UNKNOWN"

    quote_healthy = _evaluate_quote_sla(quote_age)
    # Missing news/signal data doesn't count against overall health — it's
    # structurally absent for many assets (e.g. crypto has no news signal at
    # all) rather than evidence of a real problem, so it must not drag the
    # overall status down to STALE/DEGRADED. See _dimension_status for the
    # honest per-dimension label (exposed separately, not folded in here).
    news_healthy = _evaluate_news_sla(news_age) if news_age is not None else None
    signal_healthy = _evaluate_signal_sla(signal_age) if signal_age is not None else None

    if quote_healthy and news_healthy is not False and signal_healthy is not False:
        return "HEALTHY"

    if not quote_healthy:
        return "STALE" if (quote_age and quote_age <= settings.SLA_QUOTE_MAX_AGE_SEC * 3) else "DEGRADED"

    return "STALE"


class AssetHealthService(BaseService):
    def __init__(self, health_repo: AssetHealthRepository, market_repo: MarketRepository, fundamentals_repo: AssetFundamentalsRepository):
        self.health_repo = health_repo
        self.market_repo = market_repo
        self.fundamentals_repo = fundamentals_repo

    def compute(self, asset_id: uuid.UUID) -> dict[str, Any]:
        quote = self.market_repo.get_quote_by_asset_id(asset_id)
        now = datetime.now(timezone.utc)

        fundamentals_age = None
        fundamentals = self.fundamentals_repo.get(asset_id)
        if fundamentals and fundamentals.updated_at:
            fnd_updated_at = fundamentals.updated_at
            if fnd_updated_at.tzinfo is None:
                fnd_updated_at = fnd_updated_at.replace(tzinfo=timezone.utc)
            fundamentals_age = int((now - fnd_updated_at).total_seconds())

        quote_age = None
        last_success = None
        if quote:
            updated_at = quote.updated_at
            if updated_at.tzinfo is None:
                updated_at = updated_at.replace(tzinfo=timezone.utc)
            quote_age = int((now - updated_at).total_seconds())
            last_success = updated_at

        signals = get_cached_asset_signals(str(asset_id))
        signal_age = None
        news_age = None

        if signals:
            sig_updated_at_str = signals.get("updated_at")
            if sig_updated_at_str:
                try:
                    sig_updated_at = datetime.fromisoformat(sig_updated_at_str)
                    if sig_updated_at.tzinfo is None:
                        sig_updated_at = sig_updated_at.replace(tzinfo=timezone.utc)
                    signal_age = int((now - sig_updated_at).total_seconds())
                except Exception:
                    pass

            news_ts = signals.get("news_timestamp")
            if news_ts:
                try:
                    news_time = datetime.fromtimestamp(news_ts, tz=timezone.utc)
                    news_age = int((now - news_time).total_seconds())
                except Exception:
                    pass

        if signal_age is None or news_age is None:
            snapshot = self.market_repo.get_snapshot(asset_id)
            if snapshot:
                # On a signals-cache miss, signal_age falls back to
                # AssetSnapshot.updated_at — a proxy, not a direct read of
                # when signals were actually (re)computed. It's a close
                # proxy in practice (process_asset_snapshot and
                # generate_signals run back-to-back in the same chain
                # execution, and as of the get_technical_indicators dedup
                # fix now share the same fetched indicators), but the field
                # name doesn't guarantee it's measuring signal freshness
                # specifically. See EVALUATION_MODULE_AUDIT.md 3.4.
                if signal_age is None and snapshot.updated_at:
                    snap_updated_at = snapshot.updated_at
                    if snap_updated_at.tzinfo is None:
                        snap_updated_at = snap_updated_at.replace(tzinfo=timezone.utc)
                    signal_age = int((now - snap_updated_at).total_seconds())

                if news_age is None and snapshot.payload:
                    news_ts = snapshot.payload.get("news_timestamp")
                    if news_ts:
                        try:
                            news_time = datetime.fromtimestamp(news_ts, tz=timezone.utc)
                            news_age = int((now - news_time).total_seconds())
                        except Exception:
                            pass

        status = evaluate_health_status(quote_age=quote_age, news_age=news_age, signal_age=signal_age)
        news_status = _dimension_status(news_age, _evaluate_news_sla(news_age) if news_age is not None else False)

        health = AssetHealth(
            asset_id=asset_id,
            provider_name="default",
            last_successful_ingestion=last_success,
            quote_age_seconds=quote_age,
            fundamentals_age_seconds=fundamentals_age,
            signal_age_seconds=signal_age,
            news_age_seconds=news_age,
            status=status,
            updated_at=now
        )
        updated_health = self.health_repo.upsert(health)
        self.health_repo.session.commit()

        return {
            "asset_id": str(updated_health.asset_id),
            "provider_name": updated_health.provider_name,
            "status": updated_health.status,
            "quote_age_seconds": updated_health.quote_age_seconds,
            "fundamentals_age_seconds": updated_health.fundamentals_age_seconds,
            "signal_age_seconds": updated_health.signal_age_seconds,
            "news_age_seconds": updated_health.news_age_seconds,
            "news_status": news_status,
            "updated_at": updated_health.updated_at.isoformat() if updated_health.updated_at else None
        }

File: backend/src/lib/evaluation/assetHealth.ts
import { prisma } from "../../prisma";
import { getCachedAssetSignals } from "./cache";

// Port of app/core/config.py's SLA_* settings — hardcoded at their default
// values (not overridden by env in this deployment, same convention as
// sweepStaleJobLogs.ts's STALE_JOB_TIMEOUT_SECONDS). SLA_FUNDAMENTALS_MAX_AGE_SEC
// has no Node equivalent function below because Python's own
// _evaluate_fundamentals_sla is unused dead code too — fundamentals_age is
// computed and stored on AssetHealth, but evaluate_health_status never
// takes it as an input. Matching Python's actual behavior, not its
// unused-but-defined helper.
const SLA_QUOTE_MAX_AGE_SEC = 300;
const SLA_NEWS_MAX_AGE_SEC = 3600;
const SLA_SIGNAL_MAX_AGE_SEC = 3600;

function evaluateQuoteSla(ageSeconds: number | null): boolean {
  return ageSeconds !== null && ageSeconds <= SLA_QUOTE_MAX_AGE_SEC;
}
function evaluateNewsSla(ageSeconds: number | null): boolean {
  return ageSeconds !== null && ageSeconds <= SLA_NEWS_MAX_AGE_SEC;
}
function evaluateSignalSla(ageSeconds: number | null): boolean {
  return ageSeconds !== null && ageSeconds <= SLA_SIGNAL_MAX_AGE_SEC;
}

/** Tri-state label for a single health dimension — "unknown" (no data to
 * evaluate) is distinct from "healthy"/"unhealthy" (data was evaluated). */
function dimensionStatus(ageSeconds: number | null, healthy: boolean): "unknown" | "healthy" | "unhealthy" {
  if (ageSeconds === null) return "unknown";
  return healthy ? "healthy" : "unhealthy";
}

/** Port of evaluate_health_status. Missing news/signal data doesn't count
 * against overall health — it's structurally absent for many assets (e.g.
 * crypto has no news signal at all) rather than evidence of a real problem,
 * so it must not drag the overall status down to STALE/DEGRADED (the honest
 * per-dimension label is exposed separately via dimensionStatus). */
export function evaluateHealthStatus(quoteAge: number | null, newsAge: number | null, signalAge: number | null): "HEALTHY" | "STALE" | "DEGRADED" | "UNKNOWN" {
  if (quoteAge === null && newsAge === null && signalAge === null) return "UNKNOWN";

  const quoteHealthy = evaluateQuoteSla(quoteAge);
  const newsHealthy = newsAge !== null ? evaluateNewsSla(newsAge) : null;
  const signalHealthy = signalAge !== null ? evaluateSignalSla(signalAge) : null;

  if (quoteHealthy && newsHealthy !== false && signalHealthy !== false) return "HEALTHY";

  if (!quoteHealthy) {
    return quoteAge !== null && quoteAge <= SLA_QUOTE_MAX_AGE_SEC * 3 ? "STALE" : "DEGRADED";
  }

  return "STALE";
}

export interface AssetHealthResult {
  asset_id: string;
  status: string;
  quote_age_seconds: number | null;
  fundamentals_age_seconds: number | null;
  signal_age_seconds: number | null;
  news_age_seconds: number | null;
  news_status: "unknown" | "healthy" | "unhealthy";
  updated_at: string | null;
}

/** Port of AssetHealthService.compute. */
export async function computeAssetHealthFor(assetId: string): Promise<AssetHealthResult> {
  const now = new Date();

  const [quote, fundamentals, snapshot] = await Promise.all([
    prisma.latestQuote.findFirst({ where: { assetId } }),
    prisma.assetFundamentals.findUnique({ where: { assetId } }),
    prisma.assetSnapshot.findUnique({ where: { assetId } }),
  ]);

  const fundamentalsAge = fundamentals ? Math.floor((now.getTime() - fundamentals.updatedAt.getTime()) / 1000) : null;

  let quoteAge: number | null = null;
  let lastSuccess: Date | null = null;
  if (quote) {
    quoteAge = Math.floor((now.getTime() - quote.updatedAt.getTime()) / 1000);
    lastSuccess = quote.updatedAt;
  }

  const signals = await getCachedAssetSignals<{ updated_at?: string; news_timestamp?: number }>(assetId);
  let signalAge: number | null = null;
  let newsAge: number | null = null;

  if (signals) {
    if (signals.updated_at) {
      const sigUpdatedAt = new Date(signals.updated_at);
      if (!Number.isNaN(sigUpdatedAt.getTime())) signalAge = Math.floor((now.getTime() - sigUpdatedAt.getTime()) / 1000);
    }
    if (signals.news_timestamp) {
      newsAge = Math.floor((now.getTime() - signals.news_timestamp * 1000) / 1000);
    }
  }

  if (signalAge === null || newsAge === null) {
    if (snapshot) {
      // On a signals-cache miss, signal_age falls back to
      // AssetSnapshot.updatedAt — a proxy, not a direct read of when
      // signals were actually (re)computed, but a close one in practice
      // since processAssetSnapshot and generateSignals run back-to-back.
      if (signalAge === null && snapshot.updatedAt) {
        signalAge = Math.floor((now.getTime() - snapshot.updatedAt.getTime()) / 1000);
      }
      if (newsAge === null && snapshot.payload && typeof snapshot.payload === "object") {
        const newsTs = (snapshot.payload as Record<string, unknown>).news_timestamp;
        if (typeof newsTs === "number") newsAge = Math.floor((now.getTime() - newsTs * 1000) / 1000);
      }
    }
  }

  const status = evaluateHealthStatus(quoteAge, newsAge, signalAge);
  const newsStatus = dimensionStatus(newsAge, newsAge !== null ? evaluateNewsSla(newsAge) : false);

  // provider_name is part of asset_health's composite PK (asset_id,
  // provider_name) and the upsert conflict target — wiring in the real,
  // potentially-changing LatestQuote.provider here would create a new row
  // per provider switch instead of updating the existing one. Fixed
  // placeholder key, matching Python; not surfaced in any API response.
  const providerName = "default";

  await prisma.asset_health.upsert({
    where: { asset_id_provider_name: { asset_id: assetId, provider_name: providerName } },
    create: {
      asset_id: assetId,
      provider_name: providerName,
      last_successful_ingestion: lastSuccess,
      quote_age_seconds: quoteAge,
      fundamentals_age_seconds: fundamentalsAge,
      signal_age_seconds: signalAge,
      news_age_seconds: newsAge,
      status,
      created_at: now,
      updated_at: now,
    },
    update: {
      last_successful_ingestion: lastSuccess,
      quote_age_seconds: quoteAge,
      fundamentals_age_seconds: fundamentalsAge,
      signal_age_seconds: signalAge,
      news_age_seconds: newsAge,
      status,
      updated_at: now,
    },
  });

  return {
    asset_id: assetId,
    status,
    quote_age_seconds: quoteAge,
    fundamentals_age_seconds: fundamentalsAge,
    signal_age_seconds: signalAge,
    news_age_seconds: newsAge,
    news_status: newsStatus,
    updated_at: now.toISOString(),
  };
}
